### DISCUSSION TITLE
shape plugin code that has worked in scalar_rgb variant doesn't work in llvm_ad_rgb variant, Where should i debug the code?

### DISCUSSION THREAD

**Discussion by g1r4ff3** [mitsuba-renderer/mitsuba3#898] [Q&A] [answered]
Hello, I have written a custom shape plugin code that has been compiled well and tested in scalar_rgb without errors, but I tried rendering in llvm_ad_rgb variant to speed up testing the plugin and got a runtime error in python as follows:
```
Traceback (most recent call last):
  File "...\cornell_box\cbox.py", line 13, in <module>
    scene = mi.load_file(f'./cbox.xml', spp=SPP)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: ​[xml.cpp:1140] Error while loading ".\cbox.xml" (near line 126, col 10): could not instantiate shape plugin of type "torus": jit_var_mul(r2092, r0): operands have incompatible sizes! (sizess: 1, 0)
```

I'm not used to programming in terms of SIMD, so the error message can't help.
Where should i debug the code? If you need more information for answering my quesiton, i will give any information.

**Comment by njroussel** [comment]
Hi @g1r4ff3 

If you run your code through a debugger, you should be able to step through your shape's constructor. Eventually, something in this constructor will trigger that error.
Basically one of the operands does not have any size set to it. This most likely means that you haven't correctly initialized that variable. You can use use `Type my_variable = dr::zeros<Type>(size);`, for example. 
You can also check the size of your JIT variables by calling `dr::width(my_variable)`.

**Reply by g1r4ff3** [reply]
@njroussel Sorry for delaying the reply, I'll try it on my own. Thank you!

On the way i tried to debug the code, I came across a problem that i can't get it..
Could I ask you an additional question?

**Reply by njroussel** [reply]
Sure, go ahead.

**Reply by g1r4ff3** [reply]
When variants of the system such as llvm_ad_rgb, cuda_ad_rgb were selected, Type "Float" that is also specified in mitsuba.conf is a type of float array like drjit::DiffArray<drjit::LLVMArray<float>> as i know.
While debugging the code i wrote in llvm_ad_rgb variant, a code snippet causing another exception occured, which is:
```
Caught a critical exception: jitc_var_schedule(r3555): placeholder variables are used to record computation symbolically
and cannot be scheduled for evaluation! This error message could appear for
the following reasons:

1. You are using DrJit's loop or virtual function call recording feature
   and tried to perform an operation that is not permitted in this restricted
   execution mode. Please see the documentation of recorded loops/virtual
   function calls to learn about these restrictions.

2. You are accessing a variable that was modified as part of a recorded
   loop and forgot to specify it as a loop variable. Please see the
   drjit::Loop documentation for details.
```

but i didn't use drjit::Loop or something related with that in my code by myself.
Anyway while continuing debugging in llvm_ad_rgb variant, I just wanted to check temporary variables of array type like `drjit::Array<Value, 5> coeffs` to be sure that those variables have appropriate values.
But unlike scalar_rgb variant, I can't see the contents of those variables with visual studio 2022 debugger as follows:
![coeffs_code](https://github.com/mitsuba-renderer/mitsuba3/assets/100182543/fe0ec1aa-7255-49fa-a847-7e382120f332)
![coeffs](https://github.com/mitsuba-renderer/mitsuba3/assets/100182543/3c23b8f4-9f7c-47da-9bd0-8417bcddc209)

So, is it that the contents of those are supposed to be invisible or i made mistakes?
And as an aside, if the type of the drjit::Array's element is of float array such as drjit::LLVMArray<float> or drjit::CUDAArray<float>, what size does the arrays have?
Thank you for your kind reply.

**Reply by njroussel** [reply, marked as answer]
> but i didn't use drjit::Loop or something related with that in my code by myself.

No, but you might be in a recorded loop or virtual function call. In, fact if you're writing something that is used during rendering, it most likely is in one them. Simply put, the intergrator most likely has a recorded loop so anything done inside the integrator will produce this message. For example, let's say you've written a new BSDF. This BSDF does not need a loop, but it is still called as part of the integrator's loop.

I'd recommend you read a bit more about `drjit`. Because it is a just-in-time compiler, the debugger will not behave as you expect: at any point in time you will not be able to see the value of some variable because all it is doing is recording the computation and not actually computing anything.
If you want to inspect some value, you need to evaluate/print it. You might run into the same "placeholder variable scehduled" problem because you cannot evaluate/print inside of a symbolically recorded piece of code. The recommendation is to turn off these recording features (temporarily) by using:
```python
dr.set_flag(dr.JitFlag.LoopRecord, False)
dr.set_flag(dr.JitFlag.VCallRecord, False)
```

There are many discussions thread in this repository that go a bit deeper into this topic, in addition to the ["First steps" guide of Dr.Jit](https://drjit.readthedocs.io/en/latest/firststeps-py.html).

**Reply by g1r4ff3** [reply]
I got it. I should go into ["First steps" guide of Dr.Jit](https://drjit.readthedocs.io/en/latest/firststeps-py.html) first. Thanks!
